### PR TITLE
AVX2 optimization for vec_dot_q4_3_q8_0 and refactoring

### PR DESCRIPTION
**Pull request by sw**
Apart from adding the AVX2 optimization for Q4_3, this refactors some commonly used intrinsic sequences into `inline` functions.

### CODE REVIEW

**Comment by slaren** [comment]
```
q4_3
42.94 seconds per pass - ETA 7.81 hours
prompt eval time = 54411.09 ms /   631 tokens (   86.23 ms per token) bs=512
prompt eval time = 59126.51 ms /   631 tokens (   93.70 ms per token) bs=8
       eval time = 41742.69 ms /   255 runs   (  163.70 ms per run)

q4_1
35.42 seconds per pass - ETA 6.45 hours
prompt eval time = 52762.72 ms /   631 tokens (   83.62 ms per token) bs=512
prompt eval time = 56287.43 ms /   631 tokens (   89.20 ms per token) bs=8
       eval time = 41024.48 ms /   255 runs   (  160.88 ms per run)
```

Except with perplexity the performance looks good compared to q4_1, not sure why there is a discrepancy there.

**Review by slaren — APPROVED** [review]

**Comment by ggerganov** [comment]
Before merging this: the current `Q4_3` format / implementation is not very efficient with ARM NEON:

Time per token on M1 Pro:
- `Q4_0` : `48ms`
- `Q4_1` : `55ms`
- `Q4_2` : `48ms`
- `Q4_3` : `87ms`

I want to make it close to ~50-60 ms / token.
But I think we might have to change the format if the optimization from #1083 does not work out.

Will try to optimize this withest high priority, so we can decide on the final `Q4_3` format

File: ggml.c [hunk 1]
```diff
@@ -487,6 +487,24 @@ static inline __m128i bytes_from_nibbles_16(const uint8_t * rsi)
...
+static inline float hsum_float_8(const __m256 x) {
+    __m128 res = _mm256_extractf128_ps(x, 1);
+    res = _mm_add_ps(res, _mm256_castps256_ps128(x));
+    res = _mm_add_ps(res, _mm_movehl_ps(res, res));
+    res = _mm_add_ss(res, _mm_movehdup_ps(res));
+    return _mm_cvtss_f32(res);
+}
+
+// horizontally add 8 int32_t
+static inline int hsum_int_8(const __m256i a) {
```
Comment (pubby) [inline ggml.c:500]: I think you should call this `hsum_int_32x8`. Just 8 sounds like 8-bit ints.

**Review by ggerganov — APPROVED** [review]
The `Q4_3` format seems will remain unchanged as it is on `master`, so let's merge this.
If the AVX-only path has issues we will resolve later

After merge, will try to rebase #1109 and merge it as well

**Pull request merged**